### PR TITLE
[Merged by Bors] - feat(Data/List/Forall2): Add lemmas about List.map and SublistForall₂

### PR DESCRIPTION
**Pull request by Command-Master**
Add lemmas about the interaction of List.map with SublistForall₂, analogous to `forall₂_map_left_iff` and `forall₂_map_right_iff`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Forall2.lean [hunk 1]
```diff
@@ -324,4 +324,25 @@ theorem Sublist.sublistForall₂ {l₁ l₂ : List α} (h : l₁ <+ l₂) [IsRef
...
 
+@[simp]
+theorem sublistForall₂_map_left_iff {f : γ → α} :
+    ∀ {l u}, SublistForall₂ R (map f l) u ↔ SublistForall₂ (fun c b => R (f c) b) l u := by
+  simp [sublistForall₂_iff]
+
+@[simp]
+theorem sublistForall₂_map_right_iff {f : γ → β} :
+    ∀ {l u}, SublistForall₂ R l (map f u) ↔ SublistForall₂ (fun a c => R a (f c)) l u := by
+  simp [sublistForall₂_iff]
```
Comment (eric-wieser) [inline Mathlib/Data/List/Forall2.lean:335]: This is a non-terminal simp; can you squeeze it?

File: Mathlib/Data/List/Forall2.lean [hunk 2]
```diff
@@ -324,4 +324,25 @@ theorem Sublist.sublistForall₂ {l₁ l₂ : List α} (h : l₁ <+ l₂) [IsRef
 theorem tail_sublistForall₂_self [IsRefl α Rₐ] (l : List α) : SublistForall₂ Rₐ l.tail l :=
   l.tail_sublist.sublistForall₂
 
+@[simp]
+theorem sublistForall₂_map_left_iff {f : γ → α} :
+    ∀ {l u}, SublistForall₂ R (map f l) u ↔ SublistForall₂ (fun c b => R (f c) b) l u := by
```
Comment (eric-wieser) [inline Mathlib/Data/List/Forall2.lean:329]:
```suggestion
theorem sublistForall₂_map_left_iff {f : γ → α} {l₁ : List α} {l₂ : List β} :
    SublistForall₂ R (map f l₁) l₂ ↔ SublistForall₂ (fun c b => R (f c) b) l₁ l₂ := by
```
to match the names elsewhere

**Comment by YaelDillies** [comment]
Thank!

maintainer merge

File: Mathlib/Data/List/Forall2.lean [hunk 3]
```diff
@@ -324,4 +324,24 @@ theorem Sublist.sublistForall₂ {l₁ l₂ : List α} (h : l₁ <+ l₂) [IsRef
...
+  constructor
+  · rintro ⟨l1, h1, h2⟩
+    rw [sublist_map_iff] at h2
+    obtain ⟨l', hl1, hl2⟩ := h2
+    use l'
+    rw [hl2] at h1
+    simpa [hl1] using h1
+  · rintro ⟨l1, h1, h2⟩
+    use l1.map f
+    simp [h1, h2.map]
```
Comment (Vierkantor) [inline Mathlib/Data/List/Forall2.lean:345]: Here's a slight golfing:
```suggestion
  simp only [sublistForall₂_iff]
  constructor
  · rintro ⟨l1, h1, h2⟩
    obtain ⟨l', hl1, rfl⟩ := sublist_map_iff.mp h2
    use l'
    simpa [hl1] using h1
  · rintro ⟨l1, h1, h2⟩
    use l1.map f
    simp [h1, h2.map]
```

The tricks I used:
 * replace `h2` after the `rw` with `sublist_map_iff.mp h2`
 * when we get an equality like `hl2` between a variable and an arbitrary right hand side, we can `obtain rfl := hl2` to replace the variable with the right hand side everywhere in one go.

**Review by Vierkantor — APPROVED** [review]
Thanks!

bors d+

**Pull request closed without merging**